### ISSUE TITLE
LogManager initialization: Difference in behaviour between openj9 and hotspot

### ISSUE THREAD

**Issue by smlambert** [eclipse/openj9#8237]
We observed a difference in behaviour between openj9 and hotspot when it comes to initializing logging.  Related: https://github.com/AdoptOpenJDK/openjdk-tests/issues/1463

This difference was observed when we tried to run Quarkus tests against both implementations.  Quarkus tests depend on a JBoss implementation of LogManager, some core tests check for it, assertThat(logManager).isInstanceOf(org.jboss.logmanager.LogManager.class).  

Because of the way that the Quarkus testing is setup, where the preferred LogManager (a jboss implementation) is not set up initially, but in a later pom file, we were seeing that by that time, openj9 had already loaded the default java.util.logging.LogManager, while hotspot seems to lazily initialize the jboss LogManager.  As a current workaround to this, we set that system property at JVM startup.

Raising this as a potential issue of interest for the openj9 project, as it may be that other applications make assumptions about LogManager initialization behaviour, and it may be beneficial to lazily initialize also.

Note: this is not something that is flagged as an issue during conformance testing.

Note: Think we have also seen in the past (though the tests have been subsequently updated) that some openjdk (jtreg) tests assume that logging is lazily initialized, which was causing openj9 to fail those tests.  We will have to dig to find these examples again.

**Comment by pshipton** [comment]
@keithc-ca can you please take a look, I think you are somewhat familiar with this code.

**Comment by sophia-guo** [comment]
Probably similar to https://github.com/eclipse/openj9/issues/6266, though it is related with net library.

Original test java/net/NetworkInterface/IPv4Only.java  disable IPv6 by System property 'System.setProperty("java.net.preferIPv4Stack","true");'. Test worked with hotspot and failed with openj9. Test was brittle https://github.com/eclipse/openj9/issues/6266#issuecomment-505531856 and updated with setting -Djava.net.preferIPv4Stack=true  at initial.

**Comment by keithc-ca** [comment]
@smlambert
* Can you identify the OpenJ9 build version you were using when you noticed this?
* Can you share a link to a (hopefully small) test log that shows which test I might run to understand this?

**Comment by keithc-ca** [comment]
Using
```
OPENJ9_JAVA_OPTIONS='-Xdump:java:events=load,filter=java/util/logging/LogManager$1,request=exclusive+preempt'
```
we see that the initializer of `LogManager` due to a call to `java/lang/reflect/Field.get`:
```
3XMTHREADINFO3           Java callstack:
4XESTACKTRACE                at java/lang/VMAccess.findClassOrNull(Native Method)
4XESTACKTRACE                at java/lang/ClassLoader.loadClass(ClassLoader.java:1072(Compiled Code))
4XESTACKTRACE                at java/util/logging/LogManager.<clinit>(LogManager.java:223)
4XESTACKTRACE                at jdk/internal/misc/Unsafe.ensureClassInitialized0(Native Method)
4XESTACKTRACE                at jdk/internal/misc/Unsafe.ensureClassInitialized(Unsafe.java:1446)
4XESTACKTRACE                at jdk/internal/reflect/UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
4XESTACKTRACE                at jdk/internal/reflect/ReflectionFactory.newFieldAccessor(ReflectionFactory.java:186)
4XESTACKTRACE                at java/lang/reflect/Field.acquireFieldAccessor(Field.java:1105)
4XESTACKTRACE                at java/lang/reflect/Field.getFieldAccessor(Field.java:1086)
4XESTACKTRACE                at java/lang/reflect/Field.get(Field.java:418)
4XESTACKTRACE                at com/ibm/java/lang/management/internal/LoggingMXBeanImpl.<init>(LoggingMXBeanImpl.java:94)
4XESTACKTRACE                at com/ibm/java/lang/management/internal/LoggingMXBeanImpl$1.run(LoggingMXBeanImpl.java:58)
4XESTACKTRACE                at com/ibm/java/lang/management/internal/LoggingMXBeanImpl$1.run(LoggingMXBeanImpl.java:54)
4XESTACKTRACE                at java/security/AccessController.doPrivileged(AccessController.java:678)
4XESTACKTRACE                at com/ibm/java/lang/management/internal/LoggingMXBeanImpl.createInstance(LoggingMXBeanImpl.java:65)
4XESTACKTRACE                at com/ibm/java/lang/management/internal/LoggingMXBeanImpl.<clinit>(LoggingMXBeanImpl.java:45)
4XESTACKTRACE                at java/lang/management/DefaultPlatformMBeanProvider.<clinit>(DefaultPlatformMBeanProvider.java:63)
4XESTACKTRACE                at java/lang/management/ManagementFactory$PlatformMBeanFinder.lambda$static$0(ManagementFactory.java:927)
4XESTACKTRACE                at java/lang/management/ManagementFactory$PlatformMBeanFinder$$Lambda$27/0000000000000000.run(Bytecode PC:0)
4XESTACKTRACE                at java/security/AccessController.doPrivileged(AccessController.java:873)
4XESTACKTRACE                at java/lang/management/ManagementFactory$PlatformMBeanFinder.<clinit>(ManagementFactory.java:922)
4XESTACKTRACE                at java/lang/management/ManagementFactory.getPlatformMXBean(ManagementFactory.java:684)
4XESTACKTRACE                at java/lang/management/ManagementFactory.getRuntimeMXBean(ManagementFactory.java:364)
4XESTACKTRACE                at org/apache/maven/surefire/booter/ForkedBooter.isDebugging(ForkedBooter.java:464)
4XESTACKTRACE                at org/apache/maven/surefire/booter/ForkedBooter.setupBooter(ForkedBooter.java:94)
4XESTACKTRACE                at org/apache/maven/surefire/booter/ForkedBooter.main(ForkedBooter.java:417)
```

That call is in the constructor `LoggingMXBeanImpl(Module logging)`:

```
		Class<?> managerClass = Class.forName(logging, "java.util.logging.LogManager"); //$NON-NLS-1$

		logManager_LOGGING_MXBEAN_NAME = (String) managerClass.getField("LOGGING_MXBEAN_NAME").get(null); //$NON-NLS-1$
```
I expect that deferring the call to `get(null)` until the field value is needed will fix this.

**Comment by keithc-ca** [comment]
An alternate strategy is to use `-Xtrace` which will intersperse the stack trace with normal output which may make it easier to match a failure with the corresponding trace:
```
OPENJ9_JAVA_OPTIONS='-Xtrace:print=mt,trigger=method{java/util/logging/LogManager.<clinit>,jstacktrace}'
```

**Comment by smlambert** [comment]
You can rerun this Grinder to reproduce the problem against a jdk8 build:
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/1744/testReport/  

See the 11 logging tests fail with:

Expecting:
  <java.util.logging.LogManager@2c98b391>
to be an instance of:
  <org.jboss.logmanager.LogManager>
but was instance of:
  <java.util.logging.LogManager>
	at io.quarkus.logging.LoggingTestsHelper.getHandler(LoggingTestsHelper.java:19)
	at io.quarkus.logging.AdditionalHandlersTest.additionalHandlersConfigurationTest(AdditionalHandlersTest.java:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.test.QuarkusUnitTest$1.invoke(QuarkusUnitTest.java:150)
	at ...

Rerun link:
https://ci.adoptopenjdk.net//job/Grinder/parambuild/?JDK_VERSION=8&JDK_IMPL=openj9&BUILD_LIST=external/quarkus&PLATFORM=x86-64_linux&TARGET=quarkus_java_test&SDK_RESOURCE=nightly&DOCKER_REQUIRED=true&ADOPTOPENJDK_REPO=https://github.com/smlambert/openjdk-tests.git&ADOPTOPENJDK_BRANCH=j9_8237

**Comment by keithc-ca** [comment]
In a log running quarkus tests on OpenJ9 jdk8, we see `LogManager` initialized by `com.google.inject.internal.util.Stopwatch`; it's not clear how any implementation could behave differently.
```
[2020-01-15T18:24:16.915Z] Compile and run quarkus tests
[2020-01-15T18:24:22.482Z] 18:24:21.949*0x1f7f700              mt.3        > java/util/logging/LogManager.<clinit>()V bytecode static method
[2020-01-15T18:24:22.482Z] 18:24:21.952 0x1f7f700       j9trc_aux.0        - jstacktrace:
[2020-01-15T18:24:22.482Z] 18:24:21.952 0x1f7f700       j9trc_aux.1        - [1] java.util.logging.LogManager.<clinit> (LogManager.java:145)
[2020-01-15T18:24:22.482Z] 18:24:21.952 0x1f7f700       j9trc_aux.1        - [2] java.util.logging.Logger.demandLogger (Logger.java:448)
[2020-01-15T18:24:22.482Z] 18:24:21.958 0x1f7f700       j9trc_aux.1        - [3] java.util.logging.Logger.getLogger (Logger.java:502)
[2020-01-15T18:24:22.482Z] 18:24:21.960 0x1f7f700       j9trc_aux.1        - [4] com.google.inject.internal.util.Stopwatch.<clinit> (Stopwatch.java:27)
[2020-01-15T18:24:22.482Z] 18:24:21.961 0x1f7f700       j9trc_aux.1        - [5] com.google.inject.internal.InternalInjectorCreator.<init> (InternalInjectorCreator.java:60)
[2020-01-15T18:24:22.482Z] 18:24:21.962 0x1f7f700       j9trc_aux.1        - [6] com.google.inject.Guice.createInjector (Guice.java:87)
[2020-01-15T18:24:22.482Z] 18:24:21.964 0x1f7f700       j9trc_aux.1        - [7] com.google.inject.Guice.createInjector (Guice.java:69)
[2020-01-15T18:24:22.482Z] 18:24:21.965 0x1f7f700       j9trc_aux.1        - [8] com.google.inject.Guice.createInjector (Guice.java:59)
[2020-01-15T18:24:22.482Z] 18:24:21.966 0x1f7f700       j9trc_aux.1        - [9] org.codehaus.plexus.DefaultPlexusContainer.addPlexusInjector (DefaultPlexusContainer.java:481)
[2020-01-15T18:24:22.482Z] 18:24:21.968 0x1f7f700       j9trc_aux.1        - [10] org.codehaus.plexus.DefaultPlexusContainer.<init> (DefaultPlexusContainer.java:206)
[2020-01-15T18:24:22.482Z] 18:24:21.969 0x1f7f700       j9trc_aux.1        - [11] org.apache.maven.cli.MavenCli.container (MavenCli.java:635)
[2020-01-15T18:24:22.482Z] 18:24:21.971 0x1f7f700       j9trc_aux.1        - [12] org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
[2020-01-15T18:24:22.482Z] 18:24:21.972 0x1f7f700       j9trc_aux.1        - [13] org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [14] sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [15] sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [16] sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [17] java.lang.reflect.Method.invoke (Method.java:498)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [18] org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [19] org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [20] org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [21] org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [22] sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [23] sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [24] sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [25] java.lang.reflect.Method.invoke (Method.java:498)
[2020-01-15T18:24:22.482Z] 18:24:21.973 0x1f7f700       j9trc_aux.1        - [26] org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
[2020-01-15T18:24:22.482Z] 18:24:21.974 0x1f7f700       j9trc_aux.1        - [27] org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
[2020-01-15T18:24:22.482Z] 18:24:21.974 0x1f7f700       j9trc_aux.1        - [28] org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:55)
```

**Comment by DanHeidinga** [comment]
> ... discovered that #8237 is fixed by using the latest jdk8 release. There is still an issue with jdk11, which is being fixed.

@keithc-ca Peter mentioned the LogManager issue is already resolved in JDK8 - is that using the 0.18 release builds?

**Comment by keithc-ca** [comment]
I tested with a local build this week based on the master branch, but the fix should be in the 0.18 release branch as well: I expect the problem in JDK8 was fixed by #6397.

**Comment by DanHeidinga** [comment]
Thanks for the confirmation

**Comment by pshipton** [comment]
For the record, #6397 was included in 0.16.0
